src/envoy/alts/tsi_transport_socket.cc

```diff
@@ -0,0 +1,278 @@
+    auto *message_bytes = reinterpret_cast<unsigned char *>(
+        raw_read_buffer_.linearize(message_size));
+
+    unsigned char unprotected_buffer[4096];
+    size_t unprotected_buffer_size = sizeof(unprotected_buffer);
+    tsi_result status = TSI_OK;
+
+    ENVOY_CONN_LOG(debug, "TSI: unprotecting message size: {}",
+                   callbacks_->connection(), message_size);
+
+    while (message_size > 0) {
+      size_t unprotected_buffer_size_to_send = unprotected_buffer_size;
+      size_t processed_message_size = message_size;
+      status = tsi_frame_protector_unprotect(
+          frame_protector_, message_bytes, &processed_message_size,
+          unprotected_buffer, &unprotected_buffer_size_to_send);
+      if (status != TSI_OK) break;
+      buffer.add(unprotected_buffer, unprotected_buffer_size_to_send);
+      message_bytes += processed_message_size;
+      message_size -= processed_message_size;
+      ENVOY_CONN_LOG(debug, "TSI: unprotecting message processed: {}",
+                     callbacks_->connection(), processed_message_size);
+    }
+
+    result.bytes_processed_ = raw_read_buffer_.length() - message_size;
+    raw_read_buffer_.drain(result.bytes_processed_);
+
+    ASSERT(status == TSI_OK);
+  }
+
+  ENVOY_CONN_LOG(debug, "TSI: do ead result action {} bytes {} end_stream {}",
+                 callbacks_->connection(), enumToInt(result.action_),
+                 result.bytes_processed_, result.end_stream_read_);
+  return result;
+}
+
+Network::IoResult TsiSocket::doWrite(Buffer::Instance &buffer,
+                                     bool end_stream) {
+  if (!handshake_complete_) {
+    Network::PostIoAction action = doHandshake();
+    if (action == Network::PostIoAction::Close) {
+      return {action, 0, false};
+    }
+  }
+
+  if (handshake_complete_) {
+    size_t message_size = buffer.length();
+    auto *message_bytes =
+        reinterpret_cast<unsigned char *>(buffer.linearize(message_size));
+
+    unsigned char protected_buffer[4096];
+    size_t protected_buffer_size = sizeof(protected_buffer);
+    tsi_result result = TSI_OK;
+
+    ENVOY_CONN_LOG(debug, "TSI: protecting message size: {}",
+                   callbacks_->connection(), message_size);
+
+    while (message_size > 0) {
+      size_t protected_buffer_size_to_send = protected_buffer_size;
+      size_t processed_message_size = message_size;
+      result = tsi_frame_protector_protect(
+          frame_protector_, message_bytes, &processed_message_size,
+          protected_buffer, &protected_buffer_size_to_send);
+      ASSERT(result == TSI_OK);
+      raw_write_buffer_.add(protected_buffer, protected_buffer_size_to_send);
+      message_bytes += processed_message_size;
+      message_size -= processed_message_size;
+
+      ENVOY_CONN_LOG(debug, "TSI: protecting message processed: {}",
+                     callbacks_->connection(), processed_message_size);
+
+      // Don't forget to flush.
+      if (message_size == 0) {
```

Comment:
Can we move this if () { do {} while () } out side the while loop? I think that would make it clearer. 
Maybe add ENVOY_CONN_LOG for debugging when result != TSI_OK and still_pending_size > 0?

Reply:
Done